prov/mxm/src/mlxm_cq.c

```diff
@@ -0,0 +1,228 @@
+ * SOFTWARE.
+ */
+
+#include "mlxm.h"
+
+static ssize_t mlxm_cq_readfrom(struct fid_cq *cq, void *buf, size_t len,
+                                fi_addr_t *src_addr)
+{
+    mlxm_fid_cq_t		        *fid_cq;
+    mlxm_req_t   			*mlxm_req;
+    mxm_send_req_t			*mxm_sreq;
+    mxm_recv_req_t			*mxm_rreq;
+    struct fi_context *ctx;
+    struct fi_cq_tagged_entry *cqe =
+        (struct fi_cq_tagged_entry *) buf;
+
+    fid_cq = container_of(cq, mlxm_fid_cq_t, cq);
+
+    if (fid_cq->err_q.head)
+            return -FI_EAVAIL;
+
+    mxm_progress(fid_cq->mxm_context);
+
+    if (!fid_cq->ok_q.head) {
+        return 0;
+    }
+
+
+    MLXM_CQ_DEQUEUE(fid_cq->ok_q, ctx);
+
+    mlxm_req = ctx->internal[1];
+    cqe->flags	= 0;
+    cqe->op_context	= ctx;
+
+    if ((uint64_t)(ctx->internal[3]) == FI_SEND) {
+        mxm_sreq = &mlxm_req->mxm_req.sreq;
+        assert (mxm_sreq->base.error == MXM_OK);
+        cqe->flags	|= FI_SEND;
+        cqe->len	= mxm_sreq->base.data.buffer.length;
+        cqe->data	= mxm_sreq->op.send.imm_data;
+        cqe->tag	= mxm_sreq->op.send.tag;
+
+
+        if (src_addr) {
+            /* TODO: check the size */
+            memcpy(src_addr, &(mxm_sreq->base.conn), sizeof(mxm_conn_h));
+        }
+
+    } else {
+        mxm_rreq = &mlxm_req->mxm_req.rreq;
+
+        assert(mxm_rreq->base.error != MXM_OK);
+
+
+        cqe->flags	|= FI_RECV;
+        cqe->len	= mxm_rreq->completion.actual_len;
+        cqe->data	= mxm_rreq->completion.sender_imm;
+        cqe->tag	= mxm_rreq->completion.sender_tag;
+
+        if (src_addr) {
+            /* TODO: check the size */
+            memcpy(src_addr, &(mxm_rreq->completion.source), sizeof(mxm_conn_h));
+        }
+
+    }
+
+    cqe->tag |= (((uint64_t)mlxm_req->mq_id) << 32);
+
+
+
+
+    MPOOL_RETURN(mlxm_globals.req_pool, struct mlxm_req, mlxm_req);
+    return 1;
+}
+
+static ssize_t	mlxm_cq_read(struct fid_cq *cq, void *buf, size_t len)
+{
+        return mlxm_cq_readfrom(cq, buf, len, NULL);
+}
+
+static int mlxm_cq_close(fid_t fid)
+{
+	mlxm_fid_cq_t	*fid_cq;
+	fid_cq = container_of(fid, mlxm_fid_cq_t, cq.fid);
+	/* TODO: Do I need to free whatever is in the queue? */
+
+	free(fid_cq);
+
+	return 0;
+}
+
+static ssize_t	mlxm_cq_readerr(struct fid_cq *cq, struct fi_cq_err_entry *buf,
+                                uint64_t flags)
+{
+	struct mlxm_fid_cq		*fid_cq;
+        mlxm_req_t   			*mlxm_req;
+	mxm_send_req_t			*mxm_sreq = NULL;
+	mxm_recv_req_t			*mxm_rreq = NULL;
+        struct fi_context *ctx;
+        struct fi_cq_err_entry *cqe =
+                (struct fi_cq_err_entry *) buf;
+
+        fid_cq = container_of(cq, mlxm_fid_cq_t, cq);
+
+        if (!fid_cq->err_q.head) {
+                return 0;
+        }
+
+
+
+        MLXM_CQ_DEQUEUE(fid_cq->err_q, ctx);
+
+        mlxm_req = ctx->internal[1];
+        cqe->op_context	= ctx;
+        cqe->flags	= FI_TAGGED;
+        cqe->buf = NULL;
+        if ((uint64_t)(ctx->internal[3]) == FI_SEND) {
```

Comment:
This won't work on a 32-bit platform.  I don't know if MXM supports 32-bit systems or not, but please make sure that the MXM provider does not try to build on a 32-bit system.